src/main/java/org/candlepin/subscriptions/security/IdentityHeaderAuthenticationDetailsSource.java

```diff
@@ -56,60 +52,56 @@
     private static final Logger log =
         LoggerFactory.getLogger(IdentityHeaderAuthenticationDetailsSource.class);
 
-    // NB: Right now there are just two roles.  If we expand on that, we should externalize all the
-    //  roles to an enum, serialize the header JSON to an object, and write an actual predicate method
-    //  isUserInRole
-    public static final String ORG_ADMIN_ROLE = "ORG_ADMIN";
-    public static final String INTERNAL_ROLE = "INTERNAL";
-
-    private ObjectMapper objectMapper;
     private Attributes2GrantedAuthoritiesMapper authMapper;
     private ApplicationProperties props;
+    private RbacApi rbac;
+    private String rbacApplicationName;
+    private RoleProvider roleProvider;
 
-    public IdentityHeaderAuthenticationDetailsSource(ApplicationProperties props, ObjectMapper objectMapper,
-        Attributes2GrantedAuthoritiesMapper authMapper) {
-        this.objectMapper = objectMapper;
+    public IdentityHeaderAuthenticationDetailsSource(ApplicationProperties props,
+        Attributes2GrantedAuthoritiesMapper authMapper,
+        RbacApi rbac, String rbacApplicationName) {
         this.authMapper = authMapper;
         this.props = props;
+        this.rbac = rbac;
+        this.rbacApplicationName = rbacApplicationName;
 
         if (props.isDevMode()) {
             log.info("Running in DEV mode. Security will be disabled.");
         }
+        this.roleProvider = new RoleProvider(rbacApplicationName, props.isDevMode());
     }
 
-    protected Collection<String> getUserRoles(HttpServletRequest context) {
-        ArrayList<String> userRolesList = new ArrayList<>();
-        try {
-            String identityHeader = context.getHeader(RH_IDENTITY_HEADER);
-
-            if (StringUtils.isEmpty(identityHeader)) {
-                return userRolesList;
-            }
-
-            byte[] decodedHeader = Base64.getDecoder().decode(identityHeader);
-            Map authObject = objectMapper.readValue(decodedHeader, Map.class);
-            Map identity = (Map) authObject.getOrDefault("identity", Collections.emptyMap());
-            Map user = (Map) identity.getOrDefault("user", Collections.emptyMap());
-
-            if (Boolean.TRUE.equals(user.get("is_org_admin")) || props.isDevMode()) {
-                userRolesList.add(ORG_ADMIN_ROLE);
+    protected Collection<String> getUserRoles() {
+        List<String> permissions = new LinkedList<>();
+        if (!props.isDevMode()) {
+            try {
+                // Get all permissions for the configured application name.
+                List<Access> accessList = rbac.getCurrentUserAccess(rbacApplicationName);
```

Comment:
Would it be worth adding retry functionality here (similar to what we have in UsageSnapshotProducer)?  I'm thinking that if we get random timeouts to the RBAC service that could lead to a frustrating and inconsistent experience for the user.  And more generally, what's our policy going to be if RBAC is inaccessible for whatever reason.

Reply:
When I asked RBAC folks about this, they stated we should give a 403 (default to unauthorized).

I think if the RBAC call fails in a detectable way, it might be worth capturing as a prometheus metric... Thoughts?

Reply:
Personally, I don't think that we need a retry here. If there's issues with the RBAC service I think they'd have more issues than what our app will surface. I'm not even sure that a metric would be beneficial given how often this would be an issue. But I could be convinced. :)

Reply:
I suppose rbac itself will be aware of issues like this, won't be hard to revisit in the future if we want, so I'm fine to not worry about metrics right now (plus we might get for free in our general request metrics).